src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs

```diff
@@ -383,32 +380,46 @@ public bool KeepAlive
 
         public Guid RequestTraceIdentifier => Guid.Empty;
 
-        public IAsyncResult BeginGetClientCertificate(AsyncCallback requestCallback, object state)
+        private IAsyncResult BeginGetClientCertificateCore(AsyncCallback requestCallback, object state)
         {
-            Task<X509Certificate2> getClientCertificate = new Task<X509Certificate2>(() => GetClientCertificate());
-            return TaskToApm.Begin(getClientCertificate, requestCallback, state);
+            var asyncResult = new GetClientCertificateAsyncResult(this, state, requestCallback);
+
+            // The certificate is already retrieved by the time this method is called.
+            ClientCertState = ListenerClientCertState.Completed;
+            asyncResult.InvokeCallback(GetClientCertificate());
```

Comment:
> The reason I went for this is because GetClientCertificate evaluates to a simple field access

Is it actually fast / non-blocking?  If so, makes sense, and no scheduling of a task is necessary.  If it could throw, it should still be wrapped in a try/catch and the appropriate propagation done in the catch.

> Why are there so many choices?

ctor+Start is essentially the core thing being done in all of the case: a Task is constructed and then scheduled via Start.  But that's a bit klunky.  Task.Factory.StartNew provides a factory-based approach to doing the same thing, although it's able to add a few optimizations not possible with the ctor+Start, e.g. the ctor+Start needs to apply extra synchronization to account for the possibility that two threads try to Start the same Task instance concurrently, whereas Task.Factory.StartNew doesn't expose the Task object until after it's already started, so it can skip that synchronization.  Task.Run was added later: it's a thin-wrapper on top of StartNew (and some of the overloads of Run also incorporate Unwrap) to handle the very common-cases... so Task.Run is the easiest, generally Task.Factory.StartNew is best when you need the most control (so we end up using that in most situations in corefx src), and ctor+Start is really only used in very specific and usually niche situations (e.g. where for race condition reasons you need to be very explicit about the timing of the ctor and Start, e.g. where the delegate passed in actually references the field that the resulting Task will be stored in and you need to be certain that store has finished prior to the delegate actually being scheduled so that you know the field will be initialized by the time the delegate is asynchronously invoked).